### PR TITLE
Generic authentication with pbkdf2 and JWT

### PR DESCRIPTION
**Pull request by manikandanvengatesan**

### CODE REVIEW

File: src/authHandler.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
+const hapiJWT = require('hapi-auth-jwt2');
+
+const generateAuth = require('./authGenerator').authGenerator;
```
Comment (DharanBro) [inline src/authHandler.js]: naming. generateAuth and authGenerator sounds confusing. lets have same naming

File: src/authGenerator.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        const password = request.payload[passcodeKey];
+
+        queryOptions.where[identityKey] = username;
+
+        try {
+            const record = await sequelize.models[authModel].findOne(queryOptions);
+            if (record) {
+                if (exports.verifyPassword(password, record[passcodeKey])) {
+                    const jwt = JWT.sign(
+                        { id: record.id, username: record.firstName, scope: 'admin' },
```
Comment (DharanBro) [inline src/authGenerator.js]: We should not hardcode the scope to admin. Lets have it remove till we implement authorisation

File: src/authGenerator.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                        authConfig.secret, { algorithm: 'HS256', expiresIn: '24h' },
+                    );
+                    return { token: jwt, id: record.id };
+                }
+                return Boom.unauthorized('invalid password');
+            }
+            return Boom.unauthorized('invalid username');
+        } catch (e) {
+            return Boom.badImplementation(`server error${e}`);
```
Comment (DharanBro) [inline src/authGenerator.js]: remove the ${e} since this response will be sent to the client

File: src/authHandler.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+let sequelize;
+let authConfig;
+// bring your own validation function
+const validate = async (decoded) => {
```
Comment (DharanBro) [inline src/authHandler.js:8]: Why do we need a validate method when we are already verifying with password?

File: src/authGenerator.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            if (record) {
+                if (exports.verifyPassword(password, record[passcodeKey])) {
+                    const jwt = JWT.sign(
+                        { id: record.id, username: record.firstName, scope: 'admin' },
+                        authConfig.secret, { algorithm: 'HS256', expiresIn: '24h' },
+                    );
+                    return { token: jwt, id: record.id };
+                }
+                return Boom.unauthorized('invalid password');
+            }
```
Comment (DharanBro) [inline src/authGenerator.js:31]: Instead of having nested if, can we negative returns at first?

**Review by DharanBro — CHANGES_REQUESTED** [review]

File: src/authHandler.js (hunk 4 above)
Comment (manikandanvengatesan) [reply src/authHandler.js:8]: This if for validating the token for each request

Comment (DharanBro) [reply src/authHandler.js:8]: My bad 🤦‍♂️

**Review by DharanBro — APPROVED** [review]

**Pull request merged**